R/linear_pool_sample.R

```diff
@@ -42,17 +41,55 @@ linear_pool_sample <- function(model_out_tbl, weights = NULL,
     )
     weights_col_name <- "weight"
   }
+  weight_by_cols <- colnames(weights)[colnames(weights) != weights_col_name]
   unique_weights <- unique(weights[[weights_col_name]])
 
-  if (length(unique_weights) != 1 || length(unique_provided_samples) != 1 || !is.null(n_output_samples)) {
+  if (length(unique_weights) != 1 || length(unique_provided_samples) != 1) {
     cli::cli_abort(
       "The requested ensemble calculation doesn't satisfy all conditions:
       1) {.arg model_out_tbl} contains the same number of samples from each component model,
-      2) {.arg weights} are {.val NULL} or equal for every model,
-      3) {.arg n_output_samples} = {.val NULL}"
+      2) {.arg weights} are {.val NULL} or equal for every model"
     )
   }
 
+  if (!is.null(n_output_samples)) {
+    samples_per_combo <- samples_per_combo |>
+      dplyr::left_join(weights, weight_by_cols) |>
+      dplyr::mutate(target_samples = floor(.data[[weights_col_name]] * n_output_samples))
+
+    if (any(samples_per_combo$provided_samples < samples_per_combo$target_samples)) {
+      cli::cli_abort("Requested output samples per compound unit cannot exceed the provided samples per compound unit.")
+    }
+
+    # deal with n_output_samples not divisible evenly among component models
+    actual_output_samples <- samples_per_combo |>
+      dplyr::group_by(dplyr::across(dplyr::all_of(task_id_cols))) |>
+      dplyr::summarize(target_samples = sum(.data[["target_samples"]])) |>
+      dplyr::pull(.data[["target_samples"]]) |>
+      unique()
```

Comment:
As mentioned in my previous comment, a situation like this will not arise with the code as it currently is because it will throw a validation error when not all component models provide the same number of sample forecasts for all possible combinations of task ID variables.

I had previously played around by doing the group by operation on `comp_unit_cols` instead but found that it didn't work with the way that the function is currently coded. This is because `n_output_samples` gives the number of sample forecasts to output for each unique combination of task ID variables, NOT compound task ID sets. I think it makes sense to keep this definition, because while forecasts for different values within a dependent task are related, they are still distinct forecasts (e.g. horizon 0 forecast made today for MA flu hospitalizations is distinct from that of the horizon 1 forecast)

Reply:
Code logic changed to group by `comp_unit_cols` in commit https://github.com/hubverse-org/hubEnsembles/pull/147/commits/fb2312a6c4aa6ff79f54f7fad0bd8d7947e05ba5 to be more intuitive

Reply:
Remainder samples now are calculated on a by–compound task ID set basis to avoid the bug mentioned above with changes implemented in commit https://github.com/hubverse-org/hubEnsembles/pull/147/commits/aaf04c9628f6cd8cf89995416a968df75cca02af